plugins/outputs/postgresql/tags_cache.go

```diff
@@ -0,0 +1,164 @@
+package postgresql
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/golang/groupcache/lru"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/outputs/postgresql/columns"
+	"github.com/influxdata/telegraf/plugins/outputs/postgresql/db"
+	"github.com/influxdata/telegraf/plugins/outputs/postgresql/utils"
+)
+
+const (
+	selectTagIDTemplate = "SELECT tag_id FROM %s WHERE %s"
+)
+
+// TagsCache retrieves the appropriate tagID based on the tag values
+// from the database (used only when TagsAsForeignKey property selected).
+// Also caches the LRU tagIDs
+type tagsCache interface {
+	getTagID(target *utils.TargetColumns, metric telegraf.Metric) (int, error)
+	tagsTableName(measureName string) string
+	setDb(db db.Wrapper)
+}
+
+type defTagsCache struct {
+	cache          map[string]*lru.Cache
+	tagsAsJSONb    bool
+	tagTableSuffix string
+	schema         string
+	db             db.Wrapper
+	itemsToCache   int
+}
+
+// newTagsCache returns a new implementation of the tags cache interface with LRU memoization
+func newTagsCache(numItemsInCachePerMetric int, tagsAsJSONb bool, tagTableSuffix, schema string, db db.Wrapper) tagsCache {
+	return &defTagsCache{
+		cache:          map[string]*lru.Cache{},
+		tagsAsJSONb:    tagsAsJSONb,
+		tagTableSuffix: tagTableSuffix,
+		schema:         schema,
+		db:             db,
+		itemsToCache:   numItemsInCachePerMetric,
+	}
+}
+
+func (c *defTagsCache) setDb(db db.Wrapper) {
+	c.db = db
+}
+
+// Checks the cache for the tag set of the metric, if present returns immediately.
+// Otherwise asks the database if that tag set has already been recorded.
+// If not recorded, inserts a new row to the tags table for the specific measurement.
+// Re-caches the tagID after checking the DB.
+func (c *defTagsCache) getTagID(target *utils.TargetColumns, metric telegraf.Metric) (int, error) {
+	measureName := metric.Name()
+	tags := metric.Tags()
+	cacheKey := constructCacheKey(tags)
+	tagID, isCached := c.checkTagCache(measureName, cacheKey)
+	if isCached {
+		return tagID, nil
+	}
+
+	var whereParts []string
+	var whereValues []interface{}
+	if c.tagsAsJSONb {
+		whereParts = []string{utils.QuoteIdent(columns.TagsJSONColumn) + "= $1"}
+		numTags := len(tags)
+		if numTags > 0 {
+			d, err := utils.BuildJsonb(tags)
+			if err != nil {
+				return tagID, err
+			}
+			whereValues = []interface{}{d}
+		} else {
+			whereValues = []interface{}{nil}
+		}
+	} else {
+		whereParts = make([]string, len(target.Names)-1)
+		whereValues = make([]interface{}, len(target.Names)-1)
+		whereIndex := 1
+		for columnIndex, tagName := range target.Names[1:] {
+			if val, ok := tags[tagName]; ok {
+				whereParts[columnIndex] = utils.QuoteIdent(tagName) + " = $" + strconv.Itoa(whereIndex)
+				whereValues[whereIndex-1] = val
+			} else {
+				whereParts[whereIndex-1] = tagName + " IS NULL"
```

Comment:
```suggestion
				whereParts[whereIndex-1] = utils.QuoteIdent(tagName) + " IS NULL"
```

This case needs to be sanitized as well.  It failed on Docker input when trying on `AND org.opencontainers.image.description IS NULL`.